File: 1.7.10/src/main/java/city/windmill/ingameime/gui/Widget.java
package city.windmill.ingameime.gui;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.gui.ScaledResolution;

public class Widget extends Gui {
    public int offsetX, offsetY;
    public int TextColor = 0xFF_00_00_00;
    public int Background = 0xEB_EB_EB_EB;
    public int Padding = 1;
    public int X, Y;
    public int Width, Height;
    public boolean DrawInline = true;
    protected boolean isDirty = true;

    public boolean isActive() {
        return false;
    }

    public void layout() {
        // Update Width & Height before positioning
        Width += 2 * Padding;
        Height += 2 * Padding;

        X = offsetX;
        Y = offsetY + (DrawInline ? 0 : Minecraft.getMinecraft().fontRenderer.FONT_HEIGHT);

        // Check if exceed screen
        ScaledResolution scaledresolution = new ScaledResolution(
                Minecraft.getMinecraft(),
                Minecraft.getMinecraft().displayWidth,
                Minecraft.getMinecraft().displayHeight);
        int displayHeight = scaledresolution.getScaledHeight();
        int displayWidth = scaledresolution.getScaledWidth();
        if (X + Width > displayWidth) X = Math.max(0, displayWidth - Width);
        if (Y + Height > displayHeight) Y = (DrawInline ? displayHeight : offsetY) - Height;

        isDirty = false;
    }

    public void draw() {
        drawRect(X, Y, X + Width, Y + Height, Background);
    }

    public void setPos(int x, int y) {
        if (offsetX == x && offsetY == y) return;
        offsetX = x;
        offsetY = y;
        isDirty = true;
        layout();
    }
}

File: common/src/main/kotlin/city/windmill/ingameime/client/gui/widget/Widget.kt
package city.windmill.ingameime.client.gui.widget

import net.minecraft.client.gui.Font
import net.minecraft.client.gui.GuiComponent


abstract class Widget(val font: Font) {
    var offsetX = 0
    var offsetY = 0
    var textColor = 0xFF_00_00_00.toInt()

    @Suppress("MemberVisibilityCanBePrivate")
    var backgroundColor = 0xEB_EB_EB_EB.toInt()
    open val active get() = false
    open val width get() = padding.first * 2
    open val height get() = padding.second * 2
    open val padding = 0 to 0

    fun render(mouseX: Int, mouseY: Int, delta: Float) {
        if (active)
            draw(offsetX, offsetY, mouseX, mouseY, delta)
    }
    
    open fun draw(offsetX: Int, offsetY: Int, mouseX: Int, mouseY: Int, delta: Float) {
        //Background
        GuiComponent.fill(
            offsetX,
            offsetY,
            offsetX + width,
            offsetY + height,
            backgroundColor
        )
    }
    
    fun moveTo(x: Int, y: Int) {
        offsetX = x
        offsetY = y
    }
}